Content.Server/Power/EntitySystems/ApcSystem.cs

```diff
@@ -0,0 +1,181 @@
+
+            SubscribeLocalEvent<ApcComponent, MapInitEvent>(OnApcInit);
+            SubscribeLocalEvent<ApcComponent, ChargeChangedEvent>(OnBatteryChargeChanged);
+            SubscribeLocalEvent<ApcComponent, ApcToggleMainBreakerMessage>(OnToggleMainBreaker);
+        }
+
+        // Change the APC's state only when the battery state changes, or when it's first created.
+        private void OnBatteryChargeChanged(EntityUid uid, ApcComponent component, ChargeChangedEvent args)
+        {
+            UpdateApcState(uid, component);
+        }
+
+        private void OnApcInit(EntityUid uid, ApcComponent component, MapInitEvent args)
+        {
+            UpdateApcState(uid, component);
+        }
+
+        private void OnToggleMainBreaker(EntityUid uid, ApcComponent component, ApcToggleMainBreakerMessage args)
+        {
+            TryComp<AccessReaderComponent>(uid, out var access);
+            if (args.Session.AttachedEntity == null)
+                return;
+
+            if (access == null || _accessReader.IsAllowed(access, args.Session.AttachedEntity.Value))
+            {
+                component.MainBreakerEnabled = !component.MainBreakerEnabled;
+                Comp<PowerNetworkBatteryComponent>(uid).CanDischarge = component.MainBreakerEnabled;
+
+                UpdateUIState(uid, component);
+                SoundSystem.Play(Filter.Pvs(uid), component.OnReceiveMessageSound.GetSound(), uid, AudioParams.Default.WithVolume(-2f));
+            }
+            else
+            {
+                _popupSystem.PopupCursor(Loc.GetString("apc-component-insufficient-access"),
+                    Filter.Entities(args.Session.AttachedEntity.Value));
+            }
+        }
+
+        public void UpdateApcState(EntityUid uid,
+            ApcComponent? apc=null,
+            BatteryComponent? battery=null)
+        {
+            if (!Resolve(uid, ref apc, ref battery))
+                return;
+
+            var newState = CalcChargeState(uid, apc, battery);
+            if (newState != apc.LastChargeState && apc.LastChargeStateTime + TimeSpan.FromSeconds(ApcComponent.VisualsChangeDelay) < _gameTiming.CurTime)
+            {
+                apc.LastChargeState = newState;
+                apc.LastChargeStateTime = _gameTiming.CurTime;
+
+                if (TryComp(uid, out AppearanceComponent? appearance))
+                {
+                    appearance.SetData(ApcVisuals.ChargeState, newState);
+                }
+
+                if (TryComp(uid, out SharedPointLightComponent? light))
+                {
+                    light.Color = newState switch
+                    {
+                        ApcChargeState.Lack => ApcComponent.LackColor,
+                        ApcChargeState.Charging => ApcComponent.ChargingColor,
+                        ApcChargeState.Full => ApcComponent.FullColor,
+                        _ => ApcComponent.LackColor
+                    };
+                    light.Dirty();
```

Comment:
I'm not sure on this, but after @metalgearsloth 's comments in the e-sword PR, I think lights colours also just meant to be set in the client's `ApcVisualizer`?

Reply:
you're probably right

Reply:
doing this doesn't work for reasons unknown to me. seems like it just gets bulldozed by server

Reply:
https://discord.com/channels/310555209753690112/310555209753690112/931710744439975956